docs/design/001-prometheus.md

```diff
@@ -0,0 +1,451 @@
+
+It seems that the only category of gadgets that can provide data to be interpreted as a gauge is the
+snapshotters.
+
+```yaml
+# Number of processes by namespace / pod / container
+- name: number_of_processes
+  type: gauge
+  category: snapshot
+  gadget: process
+  aggregate:
+    - k8s.namespace
+    - k8s.pod
+    - k8s.container
+
+# Number of sockets in CLOSE_WAIT state
+- name: number_of_sockets_close_wait
+  type: gauge
+  category: snapshot
+  gadget: socket
+  aggregate:
+    - k8s.namespace
+    - k8s.pod
+    - k8s.container
+  selector:
+    - "status:CLOSE_WAIT"
+```
+
+TODO: This is not totally clear how this should work since there gadget doesn't provide a stream of
+events. In this case we should execute the gadget each time prometheus scrapes the endpoint.
+
+### Histograms
+
+The histogram definition is a more complex than the previous one, hence please check the Prometheus
+documentation:
+[https://prometheus.io/docs/concepts/metric_types/#histogram](https://prometheus.io/docs/concepts/metric_types/#histogram)
+
+We'll support the same bucket configuration as described in
+[https://github.com/cloudflare/ebpf_exporter#histograms.](https://github.com/cloudflare/ebpf_exporter#histograms.)
+
+```yaml
+# DNS replies latency
+- name: dns_latency
+  type: histogram
+  category: trace
+  gadget: dns
+  field: latency
+  bucket:
+    min: 0s
+    max: 1m
+    type: exp2
+  labels:
+    - k8s.namespace
+    - k8s.pod
+  selector:
+    - "qr:R"
+```
+
+# Implementation
+
+## Gadgets supported
+
+We want to make Prometheus supported by as many gadgets as possible, however it's currently not
+possible to support all of them. The initial implementation covers these gadgets:
+
+- Tracers: counters and histograms
+- Snapshot: gauges
+
+There are some categories that we don't know if can be supported altogether, so probably we should
+also define a per gadget support:
+
+- Audit seccomp: counters
+- Profile block-io: will be nice but can require some extra work
+
+## Metrics collection
+
+The main implementation detail of this support is where to count the metrics. This proposal includes
+two approaches that are independent and that can be implemented in parallel or one after the other:
+
+- Collect metrics in user space: A very flexible and less performant solution
+- Collect metrics in eBPF: A more performant solution that should handle most common metrics
+
+### Collection in user space
+
+In this case the counting / aggregation happens on user space. This is the simplest way to collect
+metrics, but also the most expensive one. It leverages the whole functionality of our gadgets as it
+is right now. Events are still collected in eBPF and sent to user-space as they occur, where they
+are evaluated, i.e., aggregated and filtered according to the user's configuration.
+
+This option is defined to be flexible rather than performant. For critical metrics users should use
+the metrics backed by eBPF.
```

Comment:
```suggestion
This option is defined to be flexible rather than performant. For metrics with high throughput, users should use
the metrics collection backed by eBPF (see below).
```